### PR TITLE
Detect 4xx error response from firewall

### PR DESCRIPTION
**Pull request by jasongin**
When handling a 4xx error response from a tunnel management API request, check for the absence `X-Served-By` response header to detect when the error was likely caused by a firewall intercepting the request.

### CODE REVIEW

**Comment by jasongin** [comment]
I'm pretty sure this is not going to work in browsers, since the `X-Served-By` header will be hidden by CORS. I need to test that. And then we might change CORS configuration on the service to [expose](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Expose-Headers) it.

**Comment by jasongin** [comment]
This needs to wait on the CORS change in the service that adds `X-Served-By` to exposed headers.

**Comment by jasongin** [comment]
Reviewers: There appears to be a line-endings difference with `TunnelManagementClient.cs`; I suggest enabling the "Hide whitespace" option when reviewing.

image

File: cs/src/Management/TunnelManagementClient.cs
Comment (usanth) [inline cs/src/Management/TunnelManagementClient.cs:687]: Just for my own understanding, what does the X-Served-By header have to do with detecting a firewall? Is it just assumed if the returned status code is a 4xx and that this header values indicates it couldn't reach the service then we can assume that it was due to a firewall?

Comment (jasongin) [reply cs/src/Management/TunnelManagementClient.cs:687]: The tunnel service always includes the header `X-Served-By: tunnels-cluster-....` in every response. If we get a response that doesn't have that header, or has a different value), that means the response was not from the tunnel service. So the only other possibility is that the response was from a proxy or firewall that intercepted and blocked the request.

File: ts/src/management/tunnelManagementHttpClient.ts
Comment (usanth) [inline ts/src/management/tunnelManagementHttpClient.ts:840]: Is removing the null ? intentional here? Are we sure in this case that the response is not null?

**Review by jramsay — DISMISSED** [review]

**Comment by jramsay** [comment]
note: still testing before merging

File: ts/src/management/tunnelManagementHttpClient.ts
Comment (usanth) [reply ts/src/management/tunnelManagementHttpClient.ts:840]: This makes sense, thanks!

File: ts/src/management/tunnelManagementHttpClient.ts
Comment (jramsay) [inline ts/src/management/tunnelManagementHttpClient.ts:826]: I think this might need to be:

/tunnel**s**-/

Comment (jramsay) [reply ts/src/management/tunnelManagementHttpClient.ts:826]: Tested from the browser using these changes in Live Share. Before making the request I changed the authorization header to an invalid entry. This results in a 401 and hits this code path. The `x-served-by` header is as follows:  

image

Comment (jramsay) [reply ts/src/management/tunnelManagementHttpClient.ts:826]: Results in the following errorMessage:
"The tunnel request resulted in 401 status, but the request did not reach the tunnel service. This may indicate the domain 'global.rel.tunnels.api.visualstudio.com' is blocked by a firewall."

Comment (jramsay) [reply ts/src/management/tunnelManagementHttpClient.ts:826]: I can see you have it correct in the C# SDK. Wil make this change in TS.

**Review by usanth — APPROVED** [review]

**Pull request merged**
